pkg/storer/reserve.go

```diff
@@ -209,33 +209,48 @@ func (db *DB) evictBatch(ctx context.Context, batchID []byte, upToBin uint8) (er
 		default:
 		}
 
-		err := func() error {
-			db.lock.Lock(reserveUpdateLockKey)
-			defer db.lock.Unlock(reserveUpdateLockKey)
-
-			// cache evicted chunks
-			cache := func(c swarm.Chunk) {
-				if err := db.Cache().Put(ctx, c); err != nil {
-					db.logger.Error(err, "reserve cache")
+		var evicted int
+
+		err = db.reserve.IterateBatchBin(ctx, db.repo, b, batchID, func(chunk swarm.Chunk) (bool, error) {
+			err := db.Do(ctx, func(txnRepo internal.Storage) error {
+				err := db.Cache().Put(ctx, chunk)
+				if err != nil {
+					db.logger.Warning("reserve eviction cache put", "err", err)
 				}
-			}
 
-			evicted, err := db.reserve.EvictBatchBin(ctx, db, b, batchID, cache)
+				db.lock.Lock(reserveUpdateLockKey)
+				defer db.lock.Unlock(reserveUpdateLockKey)
+
+				err = db.reserve.DeleteChunk(ctx, txnRepo, chunk)
+				if err != nil {
+					return fmt.Errorf("reserve: delete chunk: %w", err)
+				}
+				return nil
+			})
 			if err != nil {
-				return fmt.Errorf("reserve.EvictBatchBin: %w", err)
+				return false, err
 			}
+			evicted++
+			return false, nil
+		})
 
-			db.logger.Info("reserve eviction", "bin", b, "evicted", evicted, "batchID", hex.EncodeToString(batchID), "size", db.reserve.Size())
+		// if there was an error, we still need to update the chunks that have already
+		// been evicted from the reserve
+		db.lock.Lock(reserveUpdateLockKey)
```

Comment:
why lock here?

Reply:
Ohh... I missed that they were changed to atomics... I thought there was a lock there which was removed. Fixed